Add new warning UI for missing debug shaders and wireframe support.

arttu-peltonen opened the pull request:
### EDIT: See [Update](https://github.com/Unity-Technologies/Graphics/pull/5458#issuecomment-914252773) for new screenshots

### Purpose of this PR
As described by case [1342315](https://fogbugz.unity3d.com/f/cases/1342315/), wireframe mode is not supported on some platforms and graphics APIs (at least Android GLES). This is a limitation with no easy fix, and Unity has not supported wireframes on these platforms in the past (see [mention in docs](https://docs.unity3d.com/2021.2/Documentation/ScriptReference/GL-wireframe.html)).

Therefore, on these platforms, we display a warning that wireframe might not work. The capability is not exposed to C# so this is the best we can do right now, in the absence of a new C# API.

A note of this behavior should also be added to the documentation when that is written (cc @oleks-k).

Editor:
![image](https://user-images.githubusercontent.com/77337829/131116437-ebe143b1-6496-463d-988f-88bc52ceacb7.png)
![image](https://user-images.githubusercontent.com/77337829/131117152-59293020-5f9d-4eaf-97b5-9ab0fdf4fa1d.png)


Runtime:
![image](https://user-images.githubusercontent.com/77337829/131116542-a638c784-5f09-4f38-823a-e4fe3e003259.png)
![image](https://user-images.githubusercontent.com/77337829/131116589-8b2d8106-4e0b-48d4-b8c7-9497abd81b71.png)

---
### Testing status
Checked that the message apppears as expected on Windows Vulkan, but not when I switch to Windows DX11.

@evelinastaniulyte are you able to verify if wireframe works correctly on iOS? The bug only mentions Android but I'd like to verify iOS too (both Metal and GLES, if we still support that).

arttu-peltonen commented on com.unity.render-pipelines.universal/Runtime/Debug/DebugDisplaySettingsRendering.cs:
```diff
@@ -129,7 +130,25 @@ internal static DebugUI.Widget CreateMapOverlaySize(DebugDisplaySettingsRenderin
+                nameAndTooltip = Strings.WireframeNotSupportedWarning,
+                isHiddenCallback = () =>
+                {
+                    switch (SystemInfo.graphicsDeviceType)
```
That's a good idea. I believe Editor itself only runs on platforms that support wireframe rendering, so build target platform is more relevant when user sees this warning in the editor.

arttu-peltonen replied:
I ended up doing something a bit different here. Instead of looking at build targets, I only show this warning in player builds. AFAIK this is never needed in the editor, only Android Vulkan/GLES have this problem. I think it's better that way you only really care about this information when you are using the debug feature on the device where it might not work correctly.

arttupeltonen commented:
Regarding Felipe's comment on changelog, I don't believe one is needed because rendering debugger is a new feature on URP.

ernestasKupciunas reviewed (approved):
Approving without testing. Good job with the PR description, all the images, and testing status! 
Code seems to be relatively straightforward and shouldn't cause any issues on the URP side.
This PR will be tested manually on mobiles by other QA.

evelinastaniulyte reviewed (changes requested):
The warning message is hardly noticeable since missing contrast in the UI. It looks like plain text.
The warning text could be set to orange color, or at least the "Warning" word could be colored and bolded.

arttu-peltonen commented:
> The warning message is hardly noticeable since missing contrast in the UI. It looks like plain text.
> The warning text could be set to orange color, or at least the "Warning" word could be colored or bolded.

@TheoWong-pixel what do you think? Might make sense, though the warning message will still look a bit clunky. Might be worth adding better support for warning labels such as this one...

TheoWong-pixel commented:
Yeah this one should be an Info box. and the tooltip should be the message inside. :)

arttu-peltonen commented:
**Update**

- Added new `DebugUI.MessageBox` widget type to allow displaying info/warning/error style messages. 
  - Ideally, this kind of UI would be tied into the widgets themselves but with current architecture I thought it easier to just make them separate widgets, and control their visibility as usual using `isHiddenCallback`.
  - Editor
  ![image](https://user-images.githubusercontent.com/77337829/132338023-012b9399-e208-477c-9cc4-7fa3a7e0a80f.png)
  - Runtime 
  ![image](https://user-images.githubusercontent.com/77337829/132338053-16cf9515-1f1b-44f9-aba8-158f92775570.png)

- Using the new MessageBox warning for the "wireframe may not be supported" message. It is only shown in player builds (Android is the only known platform to not support this).
![image](https://user-images.githubusercontent.com/77337829/132339504-6903fd53-5e5e-484c-b019-bd7c960ccfe5.png)


- Added a new warning "Debug shaders missing" to clarify situations where user opens the debug UI and most modes are seemingly broken. 
  - URP
  ![image](https://user-images.githubusercontent.com/77337829/132339447-7be53279-f693-42f7-859e-6423f77c2e36.png)
  - HDRP
  ![image](https://user-images.githubusercontent.com/77337829/132340339-2a464ce7-e3ce-44e8-992c-a3fc011b4f72.png)
  - Warning is only visible in player builds because shaders aren't stripped in the editor. 
  - Warning is not shown on "Display Stats", "Volume" or "Camera" tabs, because they don't require debug shaders.

oleks-k reviewed (approved):
Looks good.

ernestasKupciunas reviewed (approved):
Looks good. Warnings are clear and informative enough. Manually tested using URP Template scene on mobiles.

evelinastaniulyte reviewed (approved):
Looks good to me. Warnings are informative and the orange info box attracts attention, so I see no issues from UI/UX perspective.
